File: backend/app/ai/services/clustering_service.py
"""
ApeAI — Clustering Service

Implements the iterative clustering logic using vector similarity.
Supports dynamic thresholds based on feedback type.
"""

import logging
from typing import Dict, Any, List, Optional

from backend.app.ai.gemini_client import gemini_client
from backend.app.db.supabase_client import get_supabase_client
from backend.app.services.cluster_service import create_cluster, add_feedback_to_cluster

logger = logging.getLogger(__name__)

# Thresholds as requested by user
THRESHOLDS = {
    "bug": 0.90,
    "feature_request": 0.85,
    "usability": 0.85,
    "default": 0.80
}

async def classify_feedback(content: str) -> str:
    """
    Quickly classify feedback to determine the appropriate similarity threshold.
    In a full implementation, this could use Gemini, but for speed we'll use keyword heuristics
    or a very fast Gemini 2.0 Flash call.
    """
    content_lower = content.lower()
    if any(word in content_lower for word in ["bug", "error", "fail", "broken", "crash", "wrong"]):
        return "bug"
    if any(word in content_lower for word in ["feature", "add", "new", "want", "should", "could"]):
        return "feature_request"
    return "default"

async def cluster_unprocessed_feedback(user_id: Optional[str] = None) -> Dict[str, Any]:
    """
    Finds all feedback items not yet in a cluster and groups them.
    
    Step 1: Generate embeddings for any feedback that doesn't have one yet.
    Step 2: Use vector similarity to group feedback into clusters.
    """
    db = get_supabase_client()
    
    # 1. First, ensure ALL feedback items have embeddings
    from backend.app.services.embedding_service import create_embeddings_batch
    
    try:
        embed_result = await create_embeddings_batch()
        logger.info(f"📊 Embedding batch result: {embed_result}")
    except Exception as e:
        logger.error(f"Failed to create embeddings batch: {e}")
        # Continue anyway — some embeddings may already exist
    
    # 2. Get all feedback IDs that have embeddings but NO cluster link
    feedback_res = db.table("feedback").select("id, content").execute()
    links_res = db.table("cluster_feedback").select("feedback_id").execute()
    
    linked_ids = {l["feedback_id"] for l in (links_res.data or [])}
    unprocessed = [f for f in (feedback_res.data or []) if f["id"] not in linked_ids]
    
    if not unprocessed:
        return {"processed": 0, "new_clusters": 0, "linked": 0, "message": "All feedback is already clustered."}

    logger.info(f"🔄 Processing {len(unprocessed)} unclustered feedback items")
    
    # 3. Check if we have ANY embeddings at all
    embedding_check = db.table("embeddings").select("feedback_id", count="exact").execute()
    embedding_count = embedding_check.count if embedding_check.count is not None else len(embedding_check.data or [])
    
    if embedding_count == 0:
        logger.warning("⚠️ No embeddings exist yet — cannot do similarity matching. Creating standalone clusters.")
        # Fall back: create one cluster per feedback type
        return await _create_fallback_clusters(unprocessed, user_id=user_id)
    
    new_clusters = 0
    linked = 0
    
    for item in unprocessed:
        fid = item["id"]
        content = item["content"]
        
        # Determine threshold
        fb_type = await classify_feedback(content)
        threshold = THRESHOLDS.get(fb_type, THRESHOLDS["default"])
        
        # Search for similar feedback that is ALREADY in a cluster
        from backend.app.services.embedding_service import generate_embedding
        
        try:
            embedding = await generate_embedding(content)
            vector_str = f"[{','.join(str(v) for v in embedding)}]"
            
            # Find similar items
            matches = db.rpc("match_feedback", {
                "query_embedding": vector_str,
                "match_count": 5,
                "match_threshold": threshold
            }).execute()
            
            cluster_id = None
            if matches.data:
                # Check if any match belongs to a cluster
                match_ids = [m["feedback_id"] for m in matches.data if m["feedback_id"] != fid]
                if match_ids:
                    existing_links = db.table("cluster_feedback") \
                        .select("cluster_id") \
                        .in_("feedback_id", match_ids) \
                        .limit(1) \
                        .execute()
                    
                    if existing_links.data:
                        cluster_id = existing_links.data[0]["cluster_id"]
            
            if cluster_id:
                # Link to existing cluster
                sim_score = matches.data[0]["similarity"] if matches.data else 0.8
                await add_feedback_to_cluster(cluster_id, [fid], [sim_score])
                linked += 1
            else:
                # Create new cluster
                await create_cluster(
                    title=f"New {fb_type.replace('_', ' ').title()} Cluster",
                    feedback_ids=[fid],
                    confidence_score=100.0,
                    user_id=user_id
                )
                new_clusters += 1
                
        except Exception as e:
            logger.error(f"Failed to cluster feedback {fid}: {e}")
            continue

    return {
        "processed": len(unprocessed),
        "new_clusters": new_clusters,
        "linked": linked,
        "message": f"Processed {len(unprocessed)} items: {new_clusters} new clusters, {linked} linked to existing."
    }


async def _create_fallback_clusters(unprocessed: list, user_id: Optional[str] = None) -> Dict[str, Any]:
    """
    Fallback clustering when no embeddings exist.
    Groups feedback by type (bug, feature_request, etc.) and creates one cluster per type.
    """
    groups: Dict[str, list] = {}
    for item in unprocessed:
        fb_type = await classify_feedback(item["content"])
        groups.setdefault(fb_type, []).append(item["id"])
    
    new_clusters = 0
    for fb_type, fids in groups.items():
        await create_cluster(
            title=f"{fb_type.replace('_', ' ').title()} Feedback Group",
            summary=f"Auto-grouped {len(fids)} {fb_type} feedback items (fallback clustering — no embeddings available yet).",
            feedback_ids=fids,
            confidence_score=50.0,
            user_id=user_id,
        )
        new_clusters += 1
    
    return {
        "processed": len(unprocessed),
        "new_clusters": new_clusters,
        "linked": 0,
        "message": f"Created {new_clusters} clusters using keyword-based fallback grouping ({len(unprocessed)} total items)."
    }

File: backend-node/src/ai/clusteringService.ts
/**
 * ApeAI — AI Clustering Service
 *
 * Implements the iterative clustering logic using vector similarity.
 * Mirrors Python ai/services/clustering_service.py.
 */

import { getSupabaseClient } from '../db/supabase';
import { createEmbeddingsBatch, generateEmbedding } from '../services/embeddingService';
import { createCluster, addFeedbackToCluster } from '../services/clusterService';

type DbRow = Record<string, unknown>;

const THRESHOLDS: Record<string, number> = {
  bug: 0.9,
  feature_request: 0.85,
  usability: 0.85,
  default: 0.8,
};

export function classifyFeedback(content: string): string {
  const lower = content.toLowerCase();
  if (['bug', 'error', 'fail', 'broken', 'crash', 'wrong'].some((w) => lower.includes(w))) {
    return 'bug';
  }
  if (['feature', 'add', 'new', 'want', 'should', 'could'].some((w) => lower.includes(w))) {
    return 'feature_request';
  }
  return 'default';
}

export async function clusterUnprocessedFeedback(
  userId?: string,
): Promise<DbRow> {
  const db = getSupabaseClient();

  // Step 1: Ensure all feedback has embeddings
  try {
    const embedResult = await createEmbeddingsBatch();
    console.info(`📊 Embedding batch result: ${JSON.stringify(embedResult)}`);
  } catch (err) {
    console.error('Failed to create embeddings batch:', err);
    // Continue — some embeddings may already exist
  }

  // Step 2: Get unclustered feedback
  const { data: allFeedback } = await db.from('feedback').select('id, content');
  const { data: links } = await db.from('cluster_feedback').select('feedback_id');

  const linkedIds = new Set((links ?? []).map((l: DbRow) => l['feedback_id'] as string));
  const unprocessed = (allFeedback ?? []).filter(
    (f: DbRow) => !linkedIds.has(f['id'] as string),
  );

  if (unprocessed.length === 0) {
    return {
      processed: 0,
      new_clusters: 0,
      linked: 0,
      message: 'All feedback is already clustered.',
    };
  }

  console.info(`🔄 Processing ${unprocessed.length} unclustered feedback items`);

  // Step 3: Check if any embeddings exist
  const { count: embeddingCount } = await db
    .from('embeddings')
    .select('feedback_id', { count: 'exact', head: true });

  if (!embeddingCount || embeddingCount === 0) {
    console.warn('⚠️ No embeddings exist — using fallback clustering');
    return createFallbackClusters(unprocessed as DbRow[], userId);
  }

  let newClusters = 0;
  let linked = 0;

  for (const item of unprocessed as DbRow[]) {
    const fid = item['id'] as string;
    const content = item['content'] as string;

    const fbType = classifyFeedback(content);
    const threshold = THRESHOLDS[fbType] ?? THRESHOLDS['default'];

    try {
      const embedding = await generateEmbedding(content);
      const vectorStr = `[${embedding.join(',')}]`;

      const { data: matches } = await db.rpc('match_feedback', {
        query_embedding: vectorStr,
        match_count: 5,
        match_threshold: threshold,
      });

      let clusterId: string | null = null;

      if (matches && (matches as DbRow[]).length > 0) {
        const matchIds = (matches as DbRow[])
          .map((m) => m['feedback_id'] as string)
          .filter((id) => id !== fid);

        if (matchIds.length > 0) {
          const { data: existingLinks } = await db
            .from('cluster_feedback')
            .select('cluster_id')
            .in('feedback_id', matchIds)
            .limit(1);

          if (existingLinks && (existingLinks as DbRow[]).length > 0) {
            clusterId = (existingLinks as DbRow[])[0]['cluster_id'] as string;
          }
        }
      }

      if (clusterId) {
        const simScore =
          matches && (matches as DbRow[]).length > 0
            ? ((matches as DbRow[])[0]['similarity'] as number)
            : 0.8;
        await addFeedbackToCluster(clusterId, [fid], [simScore]);
        linked++;
      } else {
        await createCluster({
          title: `New ${fbType.replace('_', ' ').replace(/\b\w/g, (l) => l.toUpperCase())} Cluster`,
          feedbackIds: [fid],
          confidenceScore: 100.0,
          userId,
        });
        newClusters++;
      }
    } catch (err) {
      console.error(`Failed to cluster feedback ${fid}:`, err);
    }
  }

  return {
    processed: unprocessed.length,
    new_clusters: newClusters,
    linked,
    message: `Processed ${unprocessed.length} items: ${newClusters} new clusters, ${linked} linked to existing.`,
  };
}

async function createFallbackClusters(
  unprocessed: DbRow[],
  userId?: string,
): Promise<DbRow> {
  const groups: Record<string, string[]> = {};

  for (const item of unprocessed) {
    const fbType = classifyFeedback(item['content'] as string);
    if (!groups[fbType]) groups[fbType] = [];
    groups[fbType].push(item['id'] as string);
  }

  let newClusters = 0;
  for (const [fbType, fids] of Object.entries(groups)) {
    const label = fbType.replace('_', ' ').replace(/\b\w/g, (l) => l.toUpperCase());
    await createCluster({
      title: `${label} Feedback Group`,
      summary: `Auto-grouped ${fids.length} ${fbType} feedback items (fallback clustering — no embeddings available yet).`,
      feedbackIds: fids,
      confidenceScore: 50.0,
      userId,
    });
    newClusters++;
  }

  return {
    processed: unprocessed.length,
    new_clusters: newClusters,
    linked: 0,
    message: `Created ${newClusters} clusters using keyword-based fallback grouping (${unprocessed.length} total items).`,
  };
}
